keps/0019-package-api-versioning.md

```diff
@@ -2,64 +2,374 @@
 
 ### Goals
 
-* Allow users to install packages from a package repository that provides packages for multiple KUDO API versions
-* Allow users to install a specific version of an application and/or a specific revision of an operator
-* Reject packages that don't have a matching KUDO API version
+#### Non-functional
+
+1. Provide an *easy* experience for operator users to interact with packages and manage operators
+1. Provide a *simple* experience for operator developers to build and maintain operators and publish packages
+1. Provide a solution that is general enough to allow for future backward and forward compatibility with regards to KUDO and operator development, while opinionated enough to minimize confusion
+1. Provide a solution that maximizes the possibility for automation
+1. Follow proven industry and academia best-practices
+
+#### Functional
+
+Allow operator developers to:
+
+- Publish _packages_ to a _package registry_
+- Release operator-related features and bug fixes to operators based on a non-latest _app version_
+
+Allow operator users to:
+
+- Install packages with a specific _package version_
+- Install packages' *latest* versions
+- Query the available _package versions_ for a _package_
+- Query the available _package versions_ for a _package_ that are compatible with a specific version of Kubernetes
+- Query the available _package versions_ for a _package_ that are compatible with a specific version of the KUDO API
+- Receive operator-related features and bug fixes on operators based on a non-latest _app version_
 
 ### Non-Goals
 
-* Handle operator dependencies on specific Kubernetes versions
+- Explain or define the scenario of multiple KUDO versions with possibly multiple KUDO API versions running on a single Kubernetes cluster
+- Explain or define the planned implementation of operator<->operator dependency graphs (e.g., Kafka operator based on Kafka 2.3.1 requires a Zookeeper operator instance with version above 3.4.10)
+- Explain or define the relationship between KUDO version and KUDO API version
+- Explain or define implementations for transparent and automatic handling of multiple package registries from the KUDO CLI program
+
+## Current State
 
-## Proposal
+An `operator.yaml` file provides versions for all components describing an operator:
 
-A `operator.yaml` file already provides versions for all components describing an operator:
-  * The dependency on KUDO is provided by the `apiVersion` and `kudoVersion` fields
-  * The bundled application version is provided by the `appVersion` field
-  * The version of the operator is provided by the `version` field
-  * The version of Kubernetes is provided by the `kubernetesVersion` field
+- The _KUDO API version_ is specified in the `apiVersion` field
```

Comment:
why do we need both kudo version and api version? could you elaborate on that somewhere?

Reply:
We need a way to determine if a given operator package is API compatible with the running KUDO manager.